src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -158,6 +182,823 @@ MultiLogicalPlanCreate(Query *originalQuery, Query *queryTree)
 
 
 /*
+ * MultiSubqueryPlanTree gets the query objects and returns logical plan
+ * for subqueries.
+ *
+ * We currently have two different code paths for creating logic plan for subqueries:
+ *   (i) subquery pushdown
+ *   (ii) single relation repartition subquery
+ *
+ * In order to do create the logical plan, we follow the algorithm below:
+ *    -  If subquery pushdown planner can plan the query
+ *        -  We're done, we create the multi plan tree and return
+ *    -  Else
+ *       - If the query is not eligible for single table repartition subquery planning
+ *            - Throw the error that the subquery pushdown planner generated
+ *       - If it is eligible for single table repartition subquery planning
+ *            - Check for the errors for single table repartition subquery planning
+ *                - If no errors found, we're done. Create the multi plan and return
+ *                - If found errors, throw it
+ */
+static MultiNode *
+MultiSubqueryPlanTree(Query *originalQuery, Query *queryTree,
+					  PlannerRestrictionContext *plannerRestrictionContext)
+{
+	MultiNode *multiQueryNode = NULL;
+	DeferredErrorMessage *subqueryPushdownError = NULL;
+
+	/*
+	 * This is a generic error check that applies to both subquery pushdown
+	 * and single table repartition subquery.
+	 */
+	ErrorIfQueryNotSupported(originalQuery);
+
+	/*
+	 * In principle, we're first trying subquery pushdown planner. If it fails
+	 * to create a logical plan, continue with trying the single table
+	 * repartition subquery planning.
+	 */
+	subqueryPushdownError = DeferErrorIfContainsUnsupportedSubquery(originalQuery,
+																	plannerRestrictionContext);
+	if (!subqueryPushdownError)
+	{
+		multiQueryNode = SubqueryPushdownMultiPlanTree(originalQuery);
+	}
+	else if (subqueryPushdownError)
+	{
+		int singleRelationRepartitionSubquery =
+			SingleRelationRepartitionSubquery(originalQuery);
+
+		RangeTblEntry *subqueryRangeTableEntry = NULL;
+		Query *subqueryTree = NULL;
+		DeferredErrorMessage *repartitionQueryError = NULL;
+		List *subqueryEntryList = NULL;
+
+		/*
+		 * If not eligible for single relation repartition query, we should raise
+		 * subquery pushdown error.
+		 */
+		if (!singleRelationRepartitionSubquery)
+		{
+			RaiseDeferredErrorInternal(subqueryPushdownError, ERROR);
+		}
+
+		subqueryEntryList = SubqueryEntryList(queryTree);
+		subqueryRangeTableEntry = (RangeTblEntry *) linitial(subqueryEntryList);
+		Assert(subqueryRangeTableEntry->rtekind == RTE_SUBQUERY);
+
+		subqueryTree = subqueryRangeTableEntry->subquery;
+
+		repartitionQueryError = DeferErrorIfSubqueryNotSupported(subqueryTree);
```

Comment:
I know these names come from the previous code, it really bothered me and confused me (mostly likely I wrote both functions 😄 ) My final proposal :)
`DeferErrorIfContainsUnsupportedSubqueryPushdown` -> `DeferErrorIfUnsupportedSubqueryPushdown` (to make it shorter)
`DeferErrorIfSubqueryNotSupported` -> `DeferErrorIfUnsupportedSubqueryRepartition`

What do you think?